### DISCUSSION TITLE
How to log response headers like pino-http

### DISCUSSION THREAD

**Discussion by gabrieledarrigo** [adonisjs/core#4833] [Help]
I'm working with Adonis.js 6 and trying to log every HTTP request and response, similar to how [pino-http](https://github.com/pinojs/pino-http) does it. 
While requests are logged correctly, responses have empty headers when using `response.response.getHeaders()`.

Here are the details:

1. **Setup**: I log requests and responses using a custom middleware. To capture logs, it listens to the `finish` events on the `response` object.

2. **Problem**: Headers in the response are empty (`response.response.getHeaders()` returns `{}`). 
However, the private property `_headers` contains the expected data.

Here's the middleware code:

```ts
import { inject } from '@adonisjs/core'
import type { HttpContext } from '@adonisjs/core/http'
import { Logger, stdSerializers } from '@adonisjs/core/logger'
import type { NextFn } from '@adonisjs/core/types/http'

type LogLevel = 'info' | 'warn' | 'error' | 'debug'

@inject()
export default class HttpLoggerMiddleware {
  constructor(private readonly logger: Logger) {}

  async handle({ request, response }: HttpContext, next: NextFn): Promise<void> {
    await next()

    response.response.on('finish', () => {
      const level = this.getLogLevel(response.response.statusCode)

      this.logger[level]({
        req: stdSerializers.req(request.request),
        res: stdSerializers.res(response.response),
      })
    })
  }

  private getLogLevel(statusCode: number): LogLevel {
    if (statusCode < 400) {
      return 'info'
    }

    if (statusCode >= 400 && statusCode < 500) {
      return 'warn'
    }

    return 'error'
  }
}
```

The middleware is configured on the Server stack.
How can I correctly access response headers without relying on `_headers`?
Is there a better way to implement this logging middleware in Adonis.js?

**Comment by thetutlage** [comment]
The issue happens because we write headers on the response via the `writeHead` method. https://github.com/adonisjs/http-server/blob/7.x/src/response.ts#L616

You can work around it as follows.

```ts
response.response.on('finish', () => {
  const level = this.getLogLevel(response.response.statusCode)
  const req = stdSerializers.req(request.request)
  const res = stdSerializers.res(request.response)
  res.headers = response.getHeaders() as any

  this.logger[level]({
    req,
    res,
  })
})
```
